2.0-proposal/RELATIONS.md

```diff
@@ -0,0 +1,248 @@
+added to the global reactive context variable, where it can be accessed
+by the name of the relation (with dashes changed to underscores).  For example,
+if a charm had the following in its `metadata.yaml`:
+
+```yaml
+provides:
+  apache-website:
+    interface: apache-website
+requires:
+  db:
+    interface: mysql
+peers:
+  peers:
+    interface: myapp
+```
+
+Then the following context attributes would be available:
+
+```python
+from charms.reactive import context
+
+assert context.relations.apache_website
+assert context.relations.db
+assert context.relations.peers
+```
+
+Note that the `apache_website` instance would be of the class that was found in
+the `reactive/relations/apache_website/provides.py` file, the `db` instance
+would be from `reactive/relations/mysql/requires.py`, and the `peers` instance
+would be from `reactive/relations/myapp/peers.py`, in their respective layers.
+
+Also note that those instances would be available even if their relation is not
+related to any remote application.
+
+# The SimpleRelation API
+
+The `SimpleRelation` base class will provide a couple of automatically managed
+reactive flags, as well as providing collections for introspecting the state of
+the relation, including related applications and units, relation data provided
+by those units, and for associating reactive flags with particular related
+applications or units and filtering related applications and units by those
+flags.
+
+## Managed Flags
+
+The automatically managed flags are:
+
+  * `relations.{relation_name}.joined`
+  * `relations.{relation_name}.changed`
+
+where the `{relation_name}` is replaced at runtime with the appropriate
+relation name for that instance.  (The relation name for an instance can also
+be accessed as `relation_instance.relation_name`.)
+
+The `joined` flag will set as long as at least one remote unit is related via
+the relation, and will be updated whenever the list of related units changes so
+that any handler watching the flag will be re-invoked.
+
+The `changed` flag will be set as long as at least one remote unit is
+related and has set data on the relation, and it will be updated whenever
+either a unit updates its relation data or if a related unit goes away.  The
+interface can select only units whose relation data has changed using the
+`with_flag` filter method described below and this flag.
+
+## Application and Unit Collections
+
+The interface layer can iterate over related applications and units using the
+provided collections.  For example:
+
+```python
+class MyRelation(SimpleRelation):
+    class flags:
+        joined = 'relation.{relation_name}.joined'
+
+    @when(flags.joined)
+    def changed(self):
+        for app in self.applications:
+            for unit in app.units:
+                print('Unit {} has data: {}'.format(unit.name, unit.data))
+```
+
+Applications and units can have flags associated with them.  Flags that are
```

Comment:
@ajkavanagh
> However, there are lots of non-reactive charms out there that reactive charms still have to talk to.

Interface layers were introduced as a way to make sure that two charms actually speak the same language. If a charm isn't using interface layers then there's not much we can do..

> It sounds like you are starting to think of relations an a possible RPC mechanism between units, where as I see them as largely static data that exists on the edges of the two units, providing data that a unit can use.

I've always thought of relations as a way for charms to communicate. The [postgresql relationship](https://interface-pgsql.readthedocs.io/en/stable/requires.html) as an example:

 - *client*: I want a database with name x and permissions y. Install these extensions and let these units know that this database exists.
 - *postgres*: It's ready; here's the connection string.

You can define in a schema that there is a string value named "relid" but it's the interface code that interprets that "relid" means that it has to notify those units of the existence of the database.

That's actually one of the reasons that we switched from Chef to Juju, to be able to resolve more complex dependencies between services than a simple "insert this ip in that config file".

Aside: *I wouldn't call it RPC since you have no idea what the other side of the relationship actually does. You're requesting certain stuff from the other charm. How the charm implements that is up to that Charm.*